### DISCUSSION TITLE
Request to improve docs to callout risks of relying on thread-pool managed sync tasks for DB connections and deadlock

### DISCUSSION THREAD

**Discussion by apryor6** [fastapi/fastapi#9512] [Questions]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
# DI function for getting the DB
def get_db():
    db = SessionLocal()
    try:
        yield db
    finally:
        db.close()

# health route
@router.get("")
def health(db: Session = Depends(get_db)):
    try:
        resp = db.execute(text("SELECT 'foo' AS bar")).one()
    except Exception as e:
        logger.exception("Unable to query DB for liveness")
        raise HTTPException(status_code=status.HTTP_503_SERVICE_UNAVAILABLE, detail="Database is broken") from e

    if resp.bar != "foo":
        raise HTTPException(status_code=status.HTTP_503_SERVICE_UNAVAILABLE, detail="Database is broken")

    return {"health": "ok"}
```

### Description

This request is to update the FastAPI documentation to callout a potential pitfall with using DI to manage connection lifecycles with the usage of a connection pooler. In my architecture, there is both client-side connection pooling with SQLalchemy and server-side pooling to a PostgreSQL database through PgBouncer. This system is running within kubernetes and has many thousands of API replicas of a FastAPI application. Of note, this server side pooler is in "transaction mode", meaning that a connection is allocated for the lifecycle of a transaction and then returned to the pool - allowing in theory thousands of clients to be funneled through a much smaller number of server-side connections.

My specific problem is that my application stopped responding to its health checks, and that route simply pings the database with a `select 'foo'` type query. I found the application would stop responding to these health checks, kubernetes kills the pod, and we have an infinite unstable cycle.

The root problem is that (assuming I have understood the source code, docs, and surrounding discussions fully) FastAPI will run each of the following as a threaded task via `run_in_threadpool`:
- Solve dependencies, which yields the session
- Route invocation, where the session is used to perform a SELECT query. Without autocommit enabled, this implicitly issuse a BEGIN and opens a transaction (reserving that connection until transaction is completed)
- After response, deps are re-invoked for teardown which closes the connection and frees it, rolling back and completing the transaction at which point the connection is returned to the pool.

Thus, when many requests are inbound deadlock occurs when >= num_threads instances of the second bullet are occupying all threads in the pool and awaiting to be allocated a connection. This happens in the second not the first bullet because session creation is lazy. Behind these blocked tasks are the cleanup tasks for previous route invocations that would close and free the connection, but these cannot be scheduled due to the blocking ones awaiting those connections.

Because I am using kubernetes, it will eventually flag the pod as unhealthy and kill it, causing connections to free and give a brief moment of headroom before the issue re-arises. This became confusing because the system was "barely working, but working".

The solution is to deploy better isolation and explicit connection lifecycle management, such as using a context manager within the route as opposed to from DI. This is also discussed further in these issues.

https://github.com/tiangolo/fastapi/issues/3205
https://github.com/tiangolo/fastapi/issues/2894

While there is not a bug per se within FastAPI/Starlette, the lack of documentation articulating the potential pitfall here bit me and some others. Would it be possible to add a (possible "advanced info') callout box that notes these issues to prevent somebody encountering this sharp corner in the future?

### Operating System

Linux

### FastAPI Version

0.95.0

### Python Version

3.10

**Comment by nyte-owl** [comment]
I agree and ran into this same issue.  There is a related albeit different issue that occurs with async path operations as well.  The problem is as follows:

- setup an async pool with SQLAlchemy with N max connections
- send N requests to an endpoint served by an async path operation
- while FastAPI is awaiting database operations for those N requests (N connections checked out of the pool) send 1 more request to the endpoint (N+1 requests being served)
- on this last request, FastAPI will try to checkout a connection from the pool and block because none are available. It will block indefinitely or until the pool times out because:
  - async path operations are handled on the main thread
  - all the other awaited coroutines which are using connections cannot be cleaned up because those need to be handled on the main thread as well

To sum up:
- Main thread handling an async path operation is blocked waiting for a connection from the pool
- All connections are used/blocked and can't be freed because they need the main thread to be released
